### ISSUE TITLE
Question about the loss function

### ISSUE THREAD

**Issue by Edwardlzy** [hessamb/label-refinery#2]
Hi,

I find your idea that minimizing the KL divergence between the output distribution of the two networks very inspiring. In your paper, you derived that the KL divergence can be modified to a cross entropy loss when optimizing the label refinery network. I find this mostly effective but there is one part I don't quite understand (I'm new to deep learning so please forgive me if my question seems dumb).

Let's say I'm using a pretrained Alexnet_1 to generate labels to train Alexnet_2 from scratch. For simplicity let's consider the case of binary classification. For an example input image which Alexnet_1 generates a label [0.1, 0.9] and the Alexnet_2 is also predicting the image to be [0.1, 0.9], the cross entropy in this case is not 0 and would be backproped to update model parameters (which is not desired because the two models are generating the same output distribution if I understand the intention of label refinery correctly?) Furthermore, if the label generated by Alexnet_1 for another image is [0.5, 0.5] and the prediction by Alexnet_2 is also [0.5, 0.5], the cross entropy loss in this case is about 2 times more than the first case. With exactly the same output behavior, the loss is inconsistent.

With the first question in mind, I find it confusing to fine-tune a model from itself. For example, if I have a pretrained Alexnet and I want to improve it's accuracy using label refinery, I think I cannot use itself to generate labels for the crops to train itself because they would produce the same output and have a non-zero loss. Will this lead to better or worse accuracy?

Sorry for such a long post. Thank you in advance.

**Comment by hessamb** [comment]
Thanks!
Let's say our model outputs distribution D, and the target distribution is a *non one hot vector* distribution D'. There is no distribution D such that the cross entropy between D' and D is 0. The cross entropy is minimized where D = D', and at that point (despite a non-zero cross entropy value) the gradients are 0. And the gradient is all that matters.

Back to your example, it is true that the cross entropy between the target distribution [0.1, 0.9] and the output distribution [0.1, 0.9] is non-zero; however, that's the minimal value of cross entropy with respect to changes in the output distribution. For those values the gradient with respect to the output of the network will be 0 and the model will not get updated.

**Comment by Edwardlzy** [comment]
Thanks for the response! You are right the gradient is 0 for the cross entropy of two identical distributions. 

However, if the best the "student" can achieve is to produce the same output distribution as the "teacher" does, how is it possible to beat the "teacher"?
